Added BitBucket SCM provider

RobertNorthard opened the pull request:
**As** a developer
**I want** a BitBucket SCM provider.

Commit:
Added bitbucket scm

gem2578 committed:
Update to get Credentials from environment variable

gem2578 committed:
Update to get property from a file

gem2578 committed:
Remove temp delete of credential file and format BitBucket SCM provider.

gem2578 committed:
Use native Java net library for making HTTP requests to APIs

Commit:
Remove credential manager into load cartridge.

gem2578 committed:
Removed JSON slurper from SCMProviderHandler and updated JavaDoc.

dsingh07 commented:
@RobertNorthard Can we squash the commit which says "Removed not working part"?

RobertNorthard commented:
@dsingh07 rebased 👍

SachinKSingh28 commented on src/main/groovy/pluggable/scm/bitbucket/BitbucketRequestUtil.groovy:
```diff
@@ -0,0 +1,106 @@
+      case 200:
+        def json = jsonSlurper.parse(http.getInputStream())
+        for(int i = 0; i < json.size; i++){
+          repositoryList.add(json.values[i].name)
```
we should check for the paging in the json response as well.

RobertNorthard replied:
Good shout! I've defaulted the limit to 100. We will add this as an issue to support pagination.

SachinKSingh28 commented on src/main/groovy/pluggable/scm/bitbucket/BitbucketSCMProvider.groovy:
```diff
@@ -0,0 +1,241 @@
+          url.append(":");
+          url.append(this.scmPort);
+          url.append(this.bitbucketEndpoint);
+          if(this.bitbucketEndpoint != "/"){
```
should check for a trailing "/" and append if it doesn't exists.

Commit:
check for a trailing / and append if it doesn't exists.

Signed-off-by: Northard, Robert A <robert.a.northard@accenture.com>

Commit:
Default BitBucket API pagination to 100.

Signed-off-by: Northard, Robert A <robert.a.northard@accenture.com>

Commit:
Added example properties for bitbucket and changed the bitbucket endpoint variables

Commit:
Use secret tmp path for storing secret files.

Signed-off-by: Northard, Robert A <robert.a.northard@accenture.com>

RobertNorthard commented on src/main/groovy/pluggable/scm/bitbucket/BitbucketSCMProvider.groovy:
```diff
@@ -0,0 +1,241 @@
+
+        tempScript << "export GIT_SSH=\""+ tempDir + "/git_ssh.sh\"\n"
+        tempScript << "git clone " + this.getScmUrl(this.bitbucketUsername, this.bitbucketPassword) + repoNamespace + "/" + repoName + ".git " + tempDir + "/" + repoName + "\n"
+        def gitDir = "--git-dir=" + tempDir + "/" + repoName + "/.git"
```
BitBucket password should be ASCII encoded as it's a url. What if someone has an '@' in their password?

Commit:
Added URL encoding for passwords containing special characters

Commit:
Bitbucket unit tests.
Correct Git SCM url generation.
Remove Bitbucket REST API username/password credentials reading out of SCM provider constructor to create scm repo.

Commit:
Updated getScmUrl GroovyDoc to reflect correct Bitbuket ssh clone url.

Commit:
Added travis build file, dependent files and BitBucket Groovy doc.

Commit:
Remove property URL validation.

dsingh07 reviewed (commented):
Looks good @RobertNorthard @SachinKSingh28 

Have tested this one a fresh /C and it works as expected, in conjunction with https://github.com/Accenture/adop-platform-management/pull/30

Once the commits have been squashed, this is ready to be merged in.

dsingh07 commented:
LGTM, thanks @RobertNorthard @BuleGeorge @SachinKSingh28 @MarisBahtins @IrmantasM

SachinKSingh28 commented:
I think we need to publish the github pages as well.

RobertNorthard commented:
Awesome, thanks @SachinKSingh28. I've updated GitHub pages. The latest changes are now being reflected (https://accenture.github.io/adop-pluggable-scm/javadocs/)
